account-info/src/lib.rs

```diff
@@ -164,26 +204,46 @@ pub fn install_or_upgrade_contract(name: String) {
 /// Needs to start with `http://` or `https://`.
 #[no_mangle]
 fn set_url() {
+    let caller = utils::get_caller();
     let url: String = runtime::get_named_arg("url");
-
-    if !check_url(&url) {
-        revert(ContractError::BadUrlFormat)
-    }
-    set_key(&runtime::get_caller().to_string(), url);
+    let purse: Option<URef> = runtime::get_named_arg("purse");
+    Deposits::new().deposit_if_needed(&caller, purse);
+    Urls::new().set(&caller, &url);
 }
 
-/// Getter function for stored URLs. Returns data stored under the `public_key` argument.
+#[no_mangle]
+fn call_set_url() {
+    let url: String = runtime::get_named_arg("url");
+    let amount: Option<U512> = runtime::get_named_arg("amount");
+    let contract_address: ContractPackageHash = runtime::get_named_arg("contract_address");
+
+    let purse = amount.map(|amount| {
+        let purse = system::create_purse();
+        system::transfer_from_purse_to_purse(account::get_main_purse(), purse, amount, None)
+            .unwrap_or_revert();
+        purse
+    });
+    
+    let _: () = runtime::call_versioned_contract(contract_address, None, "set_url", runtime_args!{
+        "url" => url,
+        "purse" => purse
+    });
+} 
+
+/// Getter function for stored URLs. Returns data stored under the `account` argument.
 #[no_mangle]
 fn get_url() {
-    let public_key = runtime::get_named_arg::<PublicKey>("public_key");
-    let url = get_key::<String>(&pubkey_to_string(&public_key));
+    let account = runtime::get_named_arg::<AccountHash>("account");
+    let url = Urls::new().get(&account);
     runtime::ret(CLValue::from_t(url).unwrap_or_revert());
 }
 
 /// Function so the caller can remove their stored URL from the contract.
 #[no_mangle]
 fn delete_url() {
-    runtime::remove_key(&runtime::get_caller().to_string());
+    let caller = utils::get_caller();
+    Urls::new().delete(&caller);
+    Deposits::new().withdraw(&caller);
```

Comment:
Similarly here, if the URL was set by an admin ```Deposits::new().withdraw(&caller);``` will revert the changes with an error.